File: code/CompleteProject.m
clear;clc;

% "http://marsyas.info/downloads/datasets.html" dataset is available in this given website

notefolder=fullfile('genres','blues');
listname=dir(fullfile(notefolder,'*.wav'));
coeffs1=zeros(length(listname),186);

for k=1:length(listname)
    file_name=fullfile(notefolder,listname(k).name);
    [x, fs]=audioread(file_name);
    
    aFE=audioFeatureExtractor('SpectralDescriptorInput',"melSpectrum",'SampleRate',fs,"mfcc",true,"spectralCentroid",true,"spectralRolloffPoint",true);

    
    z=ZCR(x);%zero crossing rate
    rms=sqrt(mean(x.^2, 1));%rmse value
    features=extract(aFE,x);%mfcc and spectral features
    kovaryans=cov(features(:,1:13));
    coeffs1(k,:)=[mean(features,1) reshape(kovaryans,1,[]) z rms];
   
end
%%
notefolder=fullfile('genres','metal');
listname=dir(fullfile(notefolder,'*.wav'));
coeffs2=zeros(length(listname),186);

for k=1:length(listname)
    file_name=fullfile(notefolder,listname(k).name);
    [x, fs]=audioread(file_name);

    aFE=audioFeatureExtractor('SpectralDescriptorInput',"melSpectrum",'SampleRate',fs,"mfcc",true,"spectralCentroid",true,"spectralRolloffPoint",true);

    
    z=ZCR(x);%zero crossing rate
    rms=sqrt(mean(x.^2, 1));%rmse value
    features=extract(aFE,x);%mfcc and spectral features
    kovaryans=cov(features(:,1:13));
    coeffs2(k,:)=[mean(features,1) reshape(kovaryans,1,[]) z rms];
    
end
%%
notefolder=fullfile('genres','country');
listname=dir(fullfile(notefolder,'*.wav'));
coeffs3=zeros(length(listname),186);

for k=1:length(listname)
    file_name=fullfile(notefolder,listname(k).name);
    [x, fs]=audioread(file_name);

    aFE=audioFeatureExtractor('SpectralDescriptorInput',"melSpectrum",'SampleRate',fs,"mfcc",true,"spectralCentroid",true,"spectralRolloffPoint",true);

    
    z=ZCR(x);%zero crossing rate
    rms=sqrt(mean(x.^2, 1));%rmse value
    features=extract(aFE,x);%mfcc and spectral features
    kovaryans=cov(features(:,1:13));
    coeffs3(k,:)=[mean(features,1) reshape(kovaryans,1,[]) z rms];
    
end
%%
notefolder=fullfile('genres','pop');
listname=dir(fullfile(notefolder,'*.wav'));
coeffs4=zeros(length(listname),186);

for k=1:length(listname)
    file_name=fullfile(notefolder,listname(k).name);
    [x, fs]=audioread(file_name);

    aFE=audioFeatureExtractor('SpectralDescriptorInput',"melSpectrum",'SampleRate',fs,"mfcc",true,"spectralCentroid",true,"spectralRolloffPoint",true);

    
    z=ZCR(x);%zero crossing rate
    rms=sqrt(mean(x.^2, 1));%rmse value
    features=extract(aFE,x);%mfcc and spectral features
    kovaryans=cov(features(:,1:13));
    coeffs4(k,:)=[mean(features,1) reshape(kovaryans,1,[]) z rms];
    
end
%%
X=[coeffs1;coeffs2;coeffs3;coeffs4]; %combining feature vectors
y=[ones(size(coeffs1,1),1); ...
   2*ones(size(coeffs2,1),1); ...
   3*ones(size(coeffs3,1),1); ...
   4*ones(size(coeffs4,1),1)]; %labeling

%split into train and test
sample_count = size(X,1);
train_count = floor(0.9 * sample_count);
indices = randperm(sample_count);

Xtrain = X(indices(1:train_count),:);
Xtest = X(indices(train_count+1:end),:);
ytrain = y(indices(1:train_count),:);
ytest = y(indices(train_count+1:end),:);

%%
%OnevsAll Logistic Regression with Regularization

num_labels=4;
lambda1=1;%for regularization

[all_theta] = oneVsAll(Xtrain, ytrain, num_labels, lambda1);
%iterasyon says deitike sonu deiiyor

predLR = predictOneVsAll(all_theta, Xtest);
fprintf('\nTest Accuracy: %f\n', mean(double(predLR == ytest)) * 100);

C2=confusionmat(ytest,predLR);
cm2=confusionchart(C2);
cm2.Title='Logistic Regression';



%%
%Neural Networks

input_layer_size=186;
hidden_layer_size=20;
num_labels=4;

initial_Theta1 = randInitializeWeights(input_layer_size, hidden_layer_size);
initial_Theta2 = randInitializeWeights(hidden_layer_size, num_labels);

initial_nn_params = [initial_Theta1(:) ; initial_Theta2(:)];

options = optimset('MaxIter', 20000);%we can try different iteration number
lambda = 1;%we can try different lambda values

costFunction = @(p) nnCostFunction(p, ...
                                   input_layer_size, ...
                                   hidden_layer_size, ...
                                   num_labels, Xtrain, ytrain, lambda);
                               
[nn_params, cost] = fmincg(costFunction, initial_nn_params, options);

Theta1 = reshape(nn_params(1:hidden_layer_size * (input_layer_size + 1)), ...
                 hidden_layer_size, (input_layer_size + 1));

Theta2 = reshape(nn_params((1 + (hidden_layer_size * (input_layer_size + 1))):end), ...
                 num_labels, (hidden_layer_size + 1));


predNN=predict(Theta1,Theta2,Xtest);
fprintf('\nTest Accuracy: %f\n', mean(double(predNN == ytest)) * 100);


C1=confusionmat(ytest,predNN);
cm=confusionchart(C1);
cm.Title='Neural Network';

File: code/complete_project.py
"""Music genre classification pipeline translated from CompleteProject.m.

This script extracts audio features for four genres and trains:
1) One-vs-rest logistic regression
2) A simple neural network classifier

Expected folder structure:
    genres/
      blues/*.wav
      metal/*.wav
      country/*.wav
      pop/*.wav
"""

from __future__ import annotations

import argparse
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable

import numpy as np


@dataclass
class Dataset:
    X: np.ndarray
    y: np.ndarray


def _zcr(signal: np.ndarray) -> float:
    if signal.ndim > 1:
        signal = np.mean(signal, axis=1)
    signs = np.signbit(signal)
    return float(np.mean(signs[1:] != signs[:-1]))


def _rms(signal: np.ndarray) -> float:
    if signal.ndim > 1:
        signal = np.mean(signal, axis=1)
    return float(np.sqrt(np.mean(signal**2)))


def _extract_feature_vector(file_path: Path, n_mfcc: int = 13) -> np.ndarray:
    import librosa

    x, sr = librosa.load(file_path.as_posix(), sr=None, mono=True)

    mfcc = librosa.feature.mfcc(y=x, sr=sr, n_mfcc=n_mfcc).T
    centroid = librosa.feature.spectral_centroid(y=x, sr=sr).T
    rolloff = librosa.feature.spectral_rolloff(y=x, sr=sr).T

    features = np.hstack([mfcc, centroid, rolloff])
    cov_mfcc = np.cov(mfcc, rowvar=False).reshape(-1)

    vector = np.concatenate(
        [
            np.mean(features, axis=0),
            cov_mfcc,
            np.array([_zcr(x), _rms(x)], dtype=np.float64),
        ]
    )
    return vector


def _load_genre_features(genre_dir: Path) -> np.ndarray:
    wav_files = sorted(genre_dir.glob("*.wav"))
    if not wav_files:
        raise FileNotFoundError(f"No .wav files found in {genre_dir}")

    return np.vstack([_extract_feature_vector(w) for w in wav_files])


def build_dataset(base_dir: Path, genres: Iterable[str]) -> Dataset:
    feature_blocks: list[np.ndarray] = []
    labels: list[np.ndarray] = []

    for idx, genre in enumerate(genres, start=1):
        block = _load_genre_features(base_dir / genre)
        feature_blocks.append(block)
        labels.append(np.full((block.shape[0],), idx, dtype=np.int64))

    X = np.vstack(feature_blocks)
    y = np.concatenate(labels)
    return Dataset(X=X, y=y)


def main() -> None:
    parser = argparse.ArgumentParser(description="Train genre classifiers.")
    parser.add_argument("--genres-dir", type=Path, default=Path("genres"))
    parser.add_argument(
        "--test-size", type=float, default=0.1, help="Fraction for test split"
    )
    parser.add_argument("--seed", type=int, default=42)
    args = parser.parse_args()

    from sklearn.linear_model import LogisticRegression
    from sklearn.metrics import accuracy_score, confusion_matrix
    from sklearn.model_selection import train_test_split
    from sklearn.multiclass import OneVsRestClassifier
    from sklearn.neural_network import MLPClassifier

    genres = ["blues", "metal", "country", "pop"]
    data = build_dataset(args.genres_dir, genres)

    X_train, X_test, y_train, y_test = train_test_split(
        data.X,
        data.y,
        test_size=args.test_size,
        random_state=args.seed,
        stratify=data.y,
    )

    lr = OneVsRestClassifier(
        LogisticRegression(C=1.0, max_iter=5000, solver="lbfgs")
    )
    lr.fit(X_train, y_train)
    pred_lr = lr.predict(X_test)

    print(f"Logistic Regression Test Accuracy: {accuracy_score(y_test, pred_lr) * 100:.2f}")
    print("Logistic Regression Confusion Matrix:")
    print(confusion_matrix(y_test, pred_lr))

    nn = MLPClassifier(
        hidden_layer_sizes=(20,),
        alpha=1.0,
        max_iter=20000,
        random_state=args.seed,
    )
    nn.fit(X_train, y_train)
    pred_nn = nn.predict(X_test)

    print(f"Neural Network Test Accuracy: {accuracy_score(y_test, pred_nn) * 100:.2f}")
    print("Neural Network Confusion Matrix:")
    print(confusion_matrix(y_test, pred_nn))


if __name__ == "__main__":
    main()
